src/cplscheme/CouplingData.cpp

```diff
@@ -48,24 +48,30 @@ int CouplingData::nVertices() const
   return sample().values.size() / getDimensions();
 }
 
-Eigen::VectorXd &CouplingData::values()
+const Eigen::VectorXd &CouplingData::values() const
 {
   return sample().values;
 }
 
-const Eigen::VectorXd &CouplingData::values() const
+const Eigen::MatrixXd &CouplingData::gradients() const
 {
-  return sample().values;
+  return sample().gradients;
 }
 
-Eigen::MatrixXd &CouplingData::gradients()
+int CouplingData::gradientsRows() const
 {
-  return sample().gradients;
+  return sample().gradients.rows();
```

Comment:
At the risk of self-humiliation

```suggestion
  return _mesh->nVertices() * _data->getDimensions();
```

Reply:
You took the risk and I think it's the other way around :smirk: 
But good point. And I will also add an assertion.

Reply:
Looks like [f8d320a](https://github.com/precice/precice/pull/2203/commits/f8d320ab33ac5f3903028447e031017131b9d90d) broke some tests (e.g. `precice.CommunicationTests/SerializedStamples/DeserializeValuesAndGradients`).

Reply:
Using `getSize()` and `getMeshDimensions()` seems to do the trick. See 4f879e84ff80ae32fb677481b7ee9a1bdcad2e86.